pyro/poutine/trace_struct.py

```diff
@@ -103,18 +107,58 @@ def add_node(self, site_name, *args, **kwargs):
                 raise RuntimeError("Multiple {} sites named '{}'".format(kwargs['type'], site_name))
 
         # XXX should copy in case site gets mutated, or dont bother?
-        super(Trace, self).add_node(site_name, *args, **kwargs)
+        self.nodes[site_name] = kwargs
+        self._pred[site_name] = set()
+        self._succ[site_name] = set()
+
+    def add_edge(self, site1, site2):
+        for site in (site1, site2):
+            if site not in self.nodes:
+                self.add_node(site)
+        self._succ[site1].add(site2)
+        self._pred[site2].add(site1)
+
+    def remove_node(self, site_name):
+        self.nodes.pop(site_name)
+        for p in self._pred[site_name]:
+            self._succ[p].remove(site_name)
+        for s in self._succ[site_name]:
+            self._pred[s].remove(site_name)
+        self._pred.pop(site_name)
+        self._succ.pop(site_name)
+
+    def predecessors(self, site_name):
+        return self._pred[site_name]
+
+    def successors(self, site_name):
+        return self._succ[site_name]
 
     def copy(self):
         """
         Makes a shallow copy of self with nodes and edges preserved.
-        Identical to :meth:`networkx.DiGraph.copy`, but preserves the type
-        and the self.graph_type attribute
         """
-        trace = super(Trace, self).copy()
-        trace.__class__ = Trace
-        trace.graph_type = self.graph_type
-        return trace
+        new_tr = Trace(graph_type=self.graph_type)
+        new_tr.nodes.update(self.nodes)
+        new_tr._succ.update(self._succ)
+        new_tr._pred.update(self._pred)
+        return new_tr
+
+    def _dfs(self, site, visited):
+        if site in visited:
+            raise StopIteration
+        for s in self._succ[site]:
+            for node in self._dfs(s, visited):
+                yield node
+        visited.add(site)
+        yield site
+
+    def topsort(self, reverse=False):
```

Comment:
nit: Maybe call this `topological_sort` or `topo_sort`? At first I thought `.topsort(reverse=True)` was equivalent to `.bottomsort()` or something.